### PR TITLE
Domains: Adding new availability statuses and use new version of is-available endpoint.

### PR DESCRIPTION
**Pull request by delputnam**
This PR requires D8640-code and changes the calls to `checkDomainAvailability` to work with the new version of the `is-available` endpoint.

The new statuses and corresponding messages are all listed in `availailability-messages.js`.

To test this, apply the patch, sandbox public-api, and search for domains that match the following states:

- registered on the current site
- registered on another site owned by the same user
- registered by another user
- mapped on the same site for a domain that is transferrable
- mapped on the same site for a domain that is not transferrable
- mapped on another site owned by the same user, domain is transferrable
- mapped on another site owned by the same user, domain is not transferrable
- a domain that has a transfer in progress by the same user
- a domain that has a transfer in progress by a different user
- a domain with a TLD that isn't supported
- a domain that can be transferred
- a domain that is taken, but can't be transferred

### CODE REVIEW

**Comment by delputnam** [comment]
[Separate mappability/transfer in availability endpoint](https://trello.com/c/BHsrVxVk/35-separate-mappability-transfer-in-availability-endpoint)

File: client/lib/domains/registration/availability-messages.js [hunk 1]
```diff
@@ -14,13 +13,143 @@ import { getTld } from 'lib/domains';
...
+								href={ `/domains/manage/${ domain }/transfer/in/${ site }` }
+							/>
+						),
+					},
+				}
+			);
+			break;
+		case domainAvailability.TRANSFER_IN_PROGRESS:
+			message = translate(
+				"{{strong}}%(domain)s{{/strong}} is pending transfer and can't be conected to WordPress.com right now. " +
```
Comment (a8ci18n) [inline client/lib/domains/registration/availability-messages.js]: Hi! I've found a possible matching string that has already been translated **17** times:
`translate( '{{strong}}%(domain)s{{/strong}} is pending transfer and can't be connected to WordPress.com right now. {{a}}Learn more{{/a}}.' )` *ES Score: 9*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/lib/domains/registration/availability-messages.js [hunk 2]
```diff
@@ -14,13 +13,143 @@ import { getTld } from 'lib/domains';
...
+								href={ `/domains/manage/${ domain }/transfer/in/${ site }` }
+							/>
+						),
+					},
+				}
+			);
+			break;
+		case domainAvailability.TRANSFER_PENDING:
+			message = translate(
+				"{{strong}}%(domain)s{{/strong}} is pending transfer and can't be conected to WordPress.com right now. " +
```
Comment (a8ci18n) [inline client/lib/domains/registration/availability-messages.js]: Hi! I've found a possible matching string that has already been translated **17** times:
`translate( '{{strong}}%(domain)s{{/strong}} is pending transfer and can't be connected to WordPress.com right now. {{a}}Learn more{{/a}}.' )` *ES Score: 9*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/lib/domains/registration/availability-messages.js [hunk 3]
```diff
@@ -14,13 +13,143 @@ import { getTld } from 'lib/domains';
...
+				'{{strong}}%(domain)s{{/strong}} is already registered on your site %(site)s. Do you want to move it to this site? ' +
+					'{{a}}Yes, move it to this site.{{/a}}',
+				{
+					args: { domain, site },
+					components: {
+						strong: <strong />,
+						a: (
+							<a
+								rel="noopener noreferrer"
+								href={ `/domains/manage/${ domain }/transfer/other-site/${ domain }` }
```
Comment (aidvu) [inline client/lib/domains/registration/availability-messages.js]: There's a way how we create these links. There are helper functions for it. :)

File: client/lib/domains/registration/availability-messages.js [hunk 4]
```diff
@@ -13,14 +12,124 @@ import { translate } from 'i18n-calypso';
...
+		case domainAvailability.TRANSFER_PENDING_SAME_USER:
+			message = translate(
+				'{{strong}}%(domain)s{{/strong}} is pending transfer. {{a}}Check the transfer status{{/a}} to learn more.',
+				{
+					args: { domain },
+					components: {
+						strong: <strong />,
+						a: (
+							<a
+								target="_blank"
```
Comment (klimeryk) [inline client/lib/domains/registration/availability-messages.js]: Love this check and link in the message, but why open it in new tab?

File: client/lib/domains/registration/availability-messages.js [hunk 5]
```diff
@@ -13,14 +12,124 @@ import { translate } from 'i18n-calypso';
...
+			message = translate( '{{strong}}%(domain)s{{/strong}} is already registered on this site.', {
+				args: { domain },
+				components: {
+					strong: <strong />,
+				},
+			} );
+			break;
+		case domainAvailability.REGISTERED_OTHER_SITE_SAME_USER:
+			message = translate(
+				'{{strong}}%(domain)s{{/strong}} is already registered on your site %(site)s. Do you want to move it to this site? ' +
```
Comment (klimeryk) [inline client/lib/domains/registration/availability-messages.js]: 😍

File: client/lib/domains/registration/availability-messages.js [hunk 6]
```diff
@@ -13,14 +12,124 @@ import { translate } from 'i18n-calypso';
...
+							/>
+						),
+					},
+				}
+			);
+			break;
+		case domainAvailability.MAPPED_SAME_SITE_TRANSFERRABLE:
+			message = translate(
+				'{{strong}}%(domain)s{{/strong}} is already connected to this site, but registered somewhere else. Do you want to move ' +
+					'it from your current domain provider to WorcPress.com so you can manage the domain and the site ' +
```
Comment (klimeryk) [inline client/lib/domains/registration/availability-messages.js]: `WorcPress.com`? lowercase-d-instead-of-c-dangit! 😜

File: client/components/domains/register-domain-step/index.jsx [hunk 7]
```diff
@@ -392,32 +392,45 @@ class RegisterDomainStep extends React.Component {
 						return callback();
 					}
 
-					checkDomainAvailability( domain, ( error, result ) => {
-						const timeDiff = Date.now() - timestamp;
-						const status = result && result.status ? result.status : error;
-						const { AVAILABLE, UNKNOWN } = domainAvailability;
-						const isDomainAvailable = includes( [ AVAILABLE, UNKNOWN ], status );
+					checkDomainAvailability(
+						{ domainName: domain, blogId: this.props.selectedSite.ID },
```
Comment (klimeryk) [inline client/components/domains/register-domain-step/index.jsx]: This causes an error during signup (just search for any available domain in signup):
```
Uncaught TypeError: Cannot read property 'ID' of undefined
```

This, of course, also causes the notice for a domain that's being transferred in broken.

**Comment by klimeryk** [comment]
In signup, when I search for a domain I have pending transfer in, the `Check the transfer status` link leads to `/domains/manage/misc.blog/transfer/in/null` :(

**Comment by aidvu** [comment]
Ah, that should be only for same site. Good catch!

File: client/components/domains/register-domain-step/index.jsx [hunk 8]
```diff
@@ -392,32 +392,41 @@ class RegisterDomainStep extends React.Component {
...
-						if ( isDomainAvailable ) {
-							this.setState( { notice: null } );
-						} else {
-							this.showValidationErrorMessage( domain, status );
-						}
+					checkDomainAvailability(
+						{ domainName: domain, blogId: get( this.props, 'selectedSite.ID', null ) },
+						( error, result ) => {
+							const timeDiff = Date.now() - timestamp;
+							let status = get( result, 'status', error );
```
Comment (klimeryk) [inline client/components/domains/register-domain-step/index.jsx]: This is causing the linter to fail - it should be `const`.

**Comment by klimeryk** [comment]
If you search for a domain you already own _during signup_, you'll be presented with this notice:
screen shot 2017-12-12 at 15 58 45

It's super helpful in domain search, but in this context it's a bit confusing: what's "this site"? Maybe we can change the copy to fit both cases better? Something like `Do you want to move it to a different site?` (without the `Yes, blah` part - seems redundant)

File: client/lib/domains/registration/availability-messages.js [hunk 9]
```diff
@@ -13,14 +12,123 @@ import { translate } from 'i18n-calypso';
...
+			message = translate( '{{strong}}%(domain)s{{/strong}} is already connected to this site.', {
+				args: { domain },
+				components: {
+					strong: <strong />,
+				},
+			} );
+			break;
+		case domainAvailability.MAPPED_OTHER_SITE_SAME_USER:
+			message = translate(
+				'{{strong}}%(domain)s{{/strong}} is already connected to your site %(site)s. If you want to connect it to this site ' +
```
Comment (klimeryk) [inline client/lib/domains/registration/availability-messages.js]: If you trigger this notice during signup, you'll get this notice:
screen shot 2017-12-12 at 16 03 00

Notice the missing site and, again, a mention of "this site".

File: client/lib/domains/registration/availability-messages.js [hunk 10]
```diff
@@ -13,14 +12,123 @@ import { translate } from 'i18n-calypso';
...
+					'instead, we will be happy to help you do that. {{a}}Contact us.{{/a}}',
+				{
+					args: { domain, site },
+					components: {
+						strong: <strong />,
+						a: <a target="_blank" rel="noopener noreferrer" href={ support.CALYPSO_CONTACT } />,
```
Comment (klimeryk) [inline client/lib/domains/registration/availability-messages.js]: Probably no need to open it in a new window?

**Comment by klimeryk** [comment]
Unfortunately, it seems there's a regression? If you search for a domain already on the same site in Domain Search, you get this notice (which is OK):
screen shot 2017-12-12 at 16 24 28

But if you try the same on the mapping or transfer in step, you'll get this one:
screen shot 2017-12-12 at 16 24 12

**Comment by klimeryk** [comment]
If you search for a domain containing blacklisted string (wordpress or something else) in signup, a wrong notice is shown:
screen shot 2017-12-12 at 17 02 28
screen shot 2017-12-12 at 17 02 37

**Comment by klimeryk** [comment]
Seems it's the same for all non-mappable domains:
screen shot 2017-12-12 at 17 04 30

**Comment by aidvu** [comment]
Back end fixed. :(

File: client/lib/domains/registration/availability-messages.js [hunk 11]
```diff
@@ -13,14 +12,122 @@ import { translate } from 'i18n-calypso';
...
+								href={ paths.domainManagementTransferIn( site, domain ) }
+							/>
+						),
+					},
+				}
+			);
+			break;
+		case domainAvailability.TRANSFER_PENDING:
+			message = translate(
+				"{{strong}}%(domain)s{{/strong}} is pending transfer and can't be conected to WordPress.com right now. " +
```
Comment (klimeryk) [inline client/lib/domains/registration/availability-messages.js]: "connected" :)

File: client/lib/domains/registration/availability-messages.js [hunk 12]
```diff
@@ -13,14 +12,122 @@ import { translate } from 'i18n-calypso';
...
+								href={ paths.domainManagementTransferIn( site, domain ) }
+							/>
+						),
+					},
+				}
+			);
+			break;
+		case domainAvailability.TRANSFER_PENDING:
+			message = translate(
+				"{{strong}}%(domain)s{{/strong}} is pending transfer and can't be connected to WordPress.com right now. " +
```
Comment (a8ci18n) [inline client/lib/domains/registration/availability-messages.js]: Hi! I've found a possible matching string that has already been translated **17** times:
`translate( '{{strong}}%(domain)s{{/strong}} is pending transfer and can't be connected to WordPress.com right now. {{a}}Learn more{{/a}}.' )` *ES Score: 16*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

**Pull request merged**